drivers/gpio/gpio_pinctrl_stm32.c

```diff
@@ -0,0 +1,611 @@
+#include <gpio.h>
+
+#include <drivers/clock_control/stm32_clock_control.h>
+#include <interrupt_controller/exti_stm32.h>
+
+#include "gpio_utils.h" /* _gpio_manage_callback, _gpio_fire_callbacks */
+
+/**
+ * @brief Configuration info for one GPIO device
+ */
+struct gpio_stm32_config {
+	/** Label of pin-controller for this GPIO port */
+	const char *pinctrl_name;
+	/** Name of the GPIO bank (one of "GPIOA", ...) */
+	const char *bank_name;
+	/** Pin-controller pin number of GPIO port pin 0 */
+	u16_t pinctrl_base;
+	/** STM32Cube GPIOx */
+	GPIO_TypeDef *ll_gpio_port;
+};
+
+/**
+ * @brief Driver data for one GPIO device.
+ */
+struct gpio_stm32_data {
+	/** Pin mask to enabled INT pins generate a callback */
+	u32_t cb_pins;
+	/* User ISR callbacks */
+	sys_slist_t cb;
+	/** PINCTRL GPIO function */
+	u16_t pinctrl_function;
+	/** Pin controller that controls the GPIO port pins. */
+	struct device *pinctrl;
+	/** STM32Cube LL_SYSCFG_EXTI_PORTx */
+	u32_t ll_syscfg_exti_port;
+};
+
+/**
+ * @brief Convert external interrupt line number to SYSCFG line definiton.
+ *
+ * @param exti_line Number of external interrupt line [0..15].
+ * @return STM32CubeMX LL_SYSCFG_EXTI_LINEx.
+ */
+static inline u32_t gpio_stm32_ll_syscfg_exti_line(u8_t exti_line)
+{
+
+	const u32_t ll_syscfg_exti_line[16] = {
+		LL_SYSCFG_EXTI_LINE0,
+		LL_SYSCFG_EXTI_LINE1,
+		LL_SYSCFG_EXTI_LINE2,
+		LL_SYSCFG_EXTI_LINE3,
+		LL_SYSCFG_EXTI_LINE4,
+		LL_SYSCFG_EXTI_LINE5,
+		LL_SYSCFG_EXTI_LINE6,
+		LL_SYSCFG_EXTI_LINE7,
+		LL_SYSCFG_EXTI_LINE8,
+		LL_SYSCFG_EXTI_LINE9,
+		LL_SYSCFG_EXTI_LINE10,
+		LL_SYSCFG_EXTI_LINE11,
+		LL_SYSCFG_EXTI_LINE12,
+		LL_SYSCFG_EXTI_LINE13,
+		LL_SYSCFG_EXTI_LINE14,
+		LL_SYSCFG_EXTI_LINE15
+	};
+
+	return ll_syscfg_exti_line[exti_line];
+}
+
+/**
+ * @brief EXTI interrupt callback
+ *
+ * @param line External interrupt line (same as port pin idx).
+ * @param arg Device driver structure of GPIO device.
+ */
+static void gpio_stm32_isr(int line, void *arg)
+{
+	struct device *dev = arg;
+	struct gpio_stm32_data *data = dev->driver_data;
+
+	u32_t pin_mask = GPIO_PORT_PIN(line);
+
+	if (pin_mask & data->cb_pins) {
+		_gpio_fire_callbacks(&data->cb, dev, pin_mask);
+	}
+}
+
+/* --- SoC specific functions --- */
+
+/**
+ * @brief Connect external interrupt line to GPIO.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param exti_line Number of external interrupt line [0..15].
+ */
+static inline void gpio_stm32_syscfg_set_exti_source(struct device *dev,
+							     u8_t exti_line)
+{
+	struct gpio_stm32_data *data = dev->driver_data;
+
+#if !defined(CONFIG_SOC_SERIES_STM32F1X)
+	/* Assure system configuration register (SYSCFG) is
+	 * clocked by associated peripherial bus clock.
+	 * SYSCFG manages the external interrupt line
+	 * connection to the GPIOs (among other purposes).
+	 */
+	struct device *clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+
+	struct stm32_pclken pclken = {
+/**
+ * @code{.codegen}
```

Comment:
@b0661 : IMO, we should not allow code generation inside drivers function in this way.
Among the various categories of users, we should think about 2 specific kinds:
* The ones who'll be uncomfortable/reluctant to code generation/python snippets and that we'll reject this kind of code (and we want to keep them into Zephyr though)
* The ones who will enjoy code generation too much and will try to sneak a bit of it in every development

If we want to keep zephyr code uniform, and keep objective rules during reviews, I think we should limit code generation to systematic and redundant problematics, like device instanciation, pincrtl, api translation, ...
Aim is it could ease user's life but stays in the background and people don't have to mess with it every day.

Reply:
@erwango

You are right. This is just a workaround until clocks is defined sufficently in DTS. I‘m  working on it.